db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/JoinNode.java

```diff
@@ -378,6 +402,33 @@ public Optimizable modifyAccessPath(JBitSet outerTables) throws StandardExceptio
 
         super.modifyAccessPath(outerTables);
 
+        /* If any of the two children has chosen an index on expression, result
+         * columns have to be rebuilt because the number may change. Column
+         * mapping between parent node and this node is broken after reassigning
+         * resultColumns. For this reason, resultColumns of (grand-)parent nodes
+         * must be rebuilt until the top select node in the current query block.
+         *
+         * This logic handles nested JoinNode. If parent node is a PRN, its
+         * result columns are rebuilt in its modifyAccessPath() method. If
+         * parent is SelectNode, there is no need to rebuild its result columns
+         * but only replace their expressions.
+         */
+        if (leftResultSet.getResultColumns().isFromExprIndex() || rightResultSet.getResultColumns().isFromExprIndex()) {
+            ResultColumnList newResultColumns = (ResultColumnList) getNodeFactory()
+                    .getNode(C_NodeTypes.RESULT_COLUMN_LIST, getContextManager());
+            newResultColumns.copyOrderBySelect(resultColumns);
+
+            getNewResultColumns(resultColumns, newResultColumns, leftResultSet);
+            getNewResultColumns(resultColumns, newResultColumns, rightResultSet);
+            resultColumns = newResultColumns;
+            assert resultColumns.size() == leftResultSet.getResultColumns().size() + rightResultSet.getResultColumns().size();
+        }
+
+        // replace join predicates pushed down to right
+        if (leftResultSet.getResultColumns().isFromExprIndex()) {
+            rightResultSet.replaceIndexExpressions(leftResultSet.getResultColumns());
+        }
+
```

Comment:
I don't see a problem with the logic here, but the below case returns a wrong plan, maybe we are missing the replacement for post join conditions?
```
create table t1 (a1 int, b1 int, c1 int, d1 int);
create index idx_t1 on t1 (a1+1, b1+2);

explain select a1+1 from t2 left join t1 --splice-properties index=idx_t1
on a1+1=a2 where a1+1 is null;

Plan
----
Cursor(n=8,rows=20,updateMode=READ_ONLY (1),engine=OLTP (default))
  ->  ScrollInsensitive(n=8,totalCost=193.12,outputRows=20,outputHeapSize=160 B,partitions=1,parallelTasks=1)
    ->  ProjectRestrict(n=7,totalCost=184.72,outputRows=20,outputHeapSize=160 B,partitions=1,parallelTasks=1)
      ->  ProjectRestrict(n=6,totalCost=184.72,outputRows=20,outputHeapSize=160 B,partitions=1,parallelTasks=1,preds=[is null((A1[5:2] + 1))])
        ->  MergeSortLeftOuterJoin(n=5,totalCost=184.72,outputRows=20,outputHeapSize=160 B,partitions=1,parallelTasks=1,preds=[(A1[5:3] = A2[5:1])])
          ->  ProjectRestrict(n=4,totalCost=84.03,outputRows=20,outputHeapSize=160 B,partitions=1,parallelTasks=1)
            ->  IndexLookup(n=3,totalCost=84.03,outputRows=20,outputHeapSize=160 B,partitions=1,parallelTasks=1)
              ->  IndexScan[IDX_T1(1681)](n=2,totalCost=4.03,scannedRows=20,outputRows=20,outputHeapSize=160 B,partitions=1,parallelTasks=1,baseTable=T1(1664),preds=[(A1[5:3] = A2[5:1])])
          ->  TableScan[T2(1696)](n=0,totalCost=4.04,scannedRows=20,outputRows=20,outputHeapSize=80 B,partitions=1,parallelTasks=1)
```
the condition a1=a2 is pushed to the the index scan step(n=2),

Reply:
`IndexLookup` is there and `a1+1` is not replaced, root cause seems to be that the join node is removed after preprocessing, but all column references still refer to the join table number, making all expressions belong to table number 1. For this reason, expressions on table number 0 are also sent to table 1, making it think the index is not covering.

Not sure how to fix this issue at the moment.

Reply:
Right, post join conditions are not replaced correctly. I fixed it in the latest commit.